agent/consul/intention_endpoint.go

```diff
@@ -240,10 +240,18 @@ func (s *Intention) Match(
 	}
 
 	if rule != nil {
-		// We go through each entry and test the destination to check if it
-		// matches.
+		var authzContext acl.AuthorizerContext
+		// Go through each entry to ensure we have intention:read for the resource.
+
+		// TODO - should we do this instead of filtering the result set? This will only allow
```

Comment:
Nonblocking: but gonna ping @banks for visibility and feedback on this

Reply:
Yeah, the behavior here is odd but not insecure and for our own purposes it will work perfectly fine. Basically if you have permissions to read an intention by either the source side or destination but not both, and you attempt a `Intention.Match` rpc with your request setup to look for intentions on the side that you do not have permissions on, then you will get a permission denied.

In reality this seems fine as you can just request it for the correct side of things and the real use case for this is for Connect proxies to be able to query for intentions regarding themselves as destinations so that they can enforce them. And unless you go out of your way to mess up your ACLs, (like granting `service:write` to register the service/proxy but also having an explicit deny rule for intentions) then that destination service will have the necessary permissions.

Changing to using an acl filter on a result set instead shouldn't alter functionality in any meaningful way for any reasonable use-case. But I could argue that